### PR TITLE
Handle MPY versions for Circuitpython 7

### PR DESCRIPTION
**Pull request by Neradoc**
The MPY format has changed (from version 3 to 5) in Circuitpython 7 beta 2 so circup could not find the version of a module. Additionally Circup should help people update all their modules when switching from CP6 to CP7 or the other way around.
The format version is read from the first 2 bytes of the MPY file, `M\x03` for Circuitpython 6 and `C\x05` for Circuitpython 7 (starting with alpha 2 but there's no need to handle earlier versions of 7).

- the `__version__` of a mpy module is read differently for mpy versions 3 and 5 (fixes reading the version on 7)
- a board module that has a mpy version incompatible with the device's Circuitpython version is marked "out of date"
- a message informing the user of the incompatibility is displayed when prompting to update
- `circup list` now has a column stating the update cause (major, minor update or MPY format)
- tests added and updated to take the changes into account

I'm not completely sure about the changes to find the version, since I was unable to find the exact place in the Circuitpython code where those are generated, but it seemed to report the correct version string in all of my tests.

_PS: I changed the last line of `update` to work around a pylint "too-many-branches" error without having to do a bigger refactoring_

```
The following modules are out of date or probably need an update.
Major Updates may include breaking changes. Review before updating.
MPY Format changes from Circuitpython 6 to 7 require an update.

Module             Version  Latest  Update Reason  
------------------ -------- ------- -------------- 
adafruit_binascii  1.0.5    1.2.5   Minor Version  
adafruit_button    0.5.5    1.5.5   Major Version  
adafruit_dotstar   2.0.4    2.0.4   MPY Format     
adafruit_wsgi      1.1.4    1.1.4   MPY Format  
```

```
Found 4 module[s] needing update.
Please indicate which modules you wish to update:

Update 'adafruit_binascii'? [y/N]: y
Updated adafruit_binascii
'adafruit_button' is a Major Version update and may contain breaking changes. Do you want to update? [y/N]: y
Updated adafruit_button
WARNING: 'adafruit_dotstar': mpy format doesn't match the device's Circuitpython version. Updating is required.
Do you want to update? [y/N]: y
Updated adafruit_dotstar
WARNING: 'adafruit_wsgi': mpy format doesn't match the device's Circuitpython version. Updating is required.
Do you want to update? [y/N]: y
Updated adafruit_wsgi
```

### CODE REVIEW

**Comment by lesamouraipourpre** [comment]
The code that reads the .mpy header is:
https://github.com/adafruit/circuitpython/blob/7df5d74d951f41347502087ef851c19e76c7a2f7/py/persistentcode.c#L577-L598
```c
mp_raw_code_t *mp_raw_code_load(mp_reader_t *reader) {
    byte header[4];
    read_bytes(reader, header, sizeof(header));
    if (header[0] != 'C'
        || header[1] != MPY_VERSION
        || MPY_FEATURE_DECODE_FLAGS(header[2]) != MPY_FEATURE_FLAGS
        || header[3] > mp_small_int_bits()
        || read_uint(reader, NULL) > QSTR_WINDOW_SIZE) {
        mp_raise_MpyError(MP_ERROR_TEXT("Incompatible .mpy file. Please update all .mpy files. See http://adafru.it/mpy-update for more info."));
    }
    if (MPY_FEATURE_DECODE_ARCH(header[2]) != MP_NATIVE_ARCH_NONE) {
        byte arch = MPY_FEATURE_DECODE_ARCH(header[2]);
        if (!MPY_FEATURE_ARCH_TEST(arch)) {
            mp_raise_ValueError(MP_ERROR_TEXT("incompatible native .mpy architecture"));
        }
    }
    qstr_window_t qw;
    qw.idx = 0;
    mp_raw_code_t *rc = load_raw_code(reader, &qw);
    reader->close(reader->data);
    return rc;
}
```

I've looked over the code and it looks good, but I've not tested yet. I'll try creating a few test scenarios and report back later.

**Comment by Neradoc** [comment]
What I don't know is how to find information on how this is compiled: `__version__ = "1.2.3"`.
In 6 it compiles to `b"\x051.2.3\x0A__version__"` in other words: size, str, name_size, name.
In 7 it becomes: `b"\x0A1.2.3\x16\x16__version__"` or size*2, str, name_size*2 ?, name_size*2, name.
So I have... questions.

**Comment by lesamouraipourpre** [comment]
Ah!, I see what you mean. I'll have a look at the code and see if I can find anything. I do remember Scott or Dan saying something about better string reuse/compression within the .mpy file as a way of saving space. Maybe this is the reason.

**Comment by lesamouraipourpre** [comment]
Testing on a PyPortal Pynt

Test 1
  CircuitPython: 6 (Found device at /media/pi/CIRCUITPY, running CircuitPython 6.2.0.)
  Modules: Some Out of date
  circup list: suggests update
  circup update: updates
PASS

Test 2
  CircuitPython: 6
  Modules: None Out of date
  circup list: does nothing
  circup update: does nothing
PASS

Test 3
  CircuitPython: 6
  Modules: One version 7
  circup list: suggests update
  circup update: updates to v6 .mpy
PASS

Test 4
  CircuitPython: 6
  Modules: One current .py installed from github bundle
  circup list: does nothing
  circup update: does nothing
PASS

Test 5
  CircuitPython: 6
  Modules: One out-of-date .py installed from github bundle
  circup list: suggests update
  circup update: updates
PASS

Test 6
  CircuitPython: 7 (Found device at /media/pi/CIRCUITPY, running CircuitPython 7.0.0-alpha.2-646-g7df5d74d9.)
  Modules: All out of date v6 modules
  circup list: suggests update
  circup update: updates all modules to v7 .mpy
PASS

Test 7
  CircuitPython: 7
  Modules: Up to date
  circup list: does nothing
  circup update: does nothing
PASS

Test 8
  CircuitPython: 7
  Modules: One out-of-date .py installed from github bundle
  circup list: suggests update
  circup update: updates .py
PASS

Test 9
  CircuitPython: 7
  Modules: One up-to-date .py installed from PyPI
  circup freeze: reports version as 0.0.0-auto.0 (This is an already reported issue with the build chain)
  circup list: suggests update
  circup update: updates .py from github
PASS

Test 10
  CircuitPython: 7
  Modules: One v7 .mpy out-of-date
  circup list: suggests update
  circup update: goes weird
```
circup list
Found device at /media/pi/CIRCUITPY, running CircuitPython 7.0.0-alpha.2-646-g7df5d74d9.
The following modules are out of date or probably need an update.
Major Updates may include breaking changes. Review before updating.
MPY Format changes from Circuitpython 6 to 7 require an update.

Module           Version  Latest  Update Reason  
---------------- -------- ------- -------------- 
adafruit_seesaw  1.8.0    1.8.2   Minor Version  

circup on  nera-mpy7-handling via 🐍 v3.7.3 (.env) took 2s 
❯ circup update
Found device at /media/pi/CIRCUITPY, running CircuitPython 7.0.0-alpha.2-646-g7df5d74d9.
Found 1 module[s] needing update.
Please indicate which modules you wish to update:

Update 'adafruit_seesaw'? [y/N]: y
Updated adafruit_seesaw

circup on  nera-mpy7-handling via 🐍 v3.7.3 (.env) took 8s 
❯ circup list
Found device at /media/pi/CIRCUITPY, running CircuitPython 7.0.0-alpha.2-646-g7df5d74d9.
The following modules are out of date or probably need an update.
Major Updates may include breaking changes. Review before updating.
MPY Format changes from Circuitpython 6 to 7 require an update.

Module           Version  Latest  Update Reason  
---------------- -------- ------- -------------- 
adafruit_seesaw  1.2.3    1.8.2   Minor Version  
```
FAIL - The identified version of seesaw is wrong.
After further investigation the file adafruit_seesaw/neopixel.py did not have it's version set to 0.0.0-auto.0 on github. PR sent.

Retest with adafruit-bme280
  circup list: suggests update
  circup update: updates
PASS

I think I've covered most variants of upgrading to v7 and downgrading to v6.

**Comment by Neradoc** [comment]
> Test 10
> CircuitPython: 7
> Modules: One v7 .mpy out-of-date
> circup list: suggests update
> circup update: goes weird

Oh interesting, I couldn't reproduce at first. Circup takes the first version it finds in any file inside a package directory. They are most likely explored in an arbitrary order due to file system shenaningans. Not much we can do outside of fixing the repo.

**Comment by lesamouraipourpre** [comment]
> > Test 10
> > CircuitPython: 7
> > Modules: One v7 .mpy out-of-date
> > circup list: suggests update
> > circup update: goes weird
> 
> Oh interesting, I couldn't reproduce at first. Circup takes the first version it finds in any file inside a package directory. They are most likely explored in an arbitrary order due to file system shenanigans. Not much we can do outside of fixing the repo.

That file was the only one in the seesaw package that didn't have the correct version. `git blame` said had been that way for 3 years. Anyway, Ladyada has already merged the PR. Just waiting for a release.

**Comment by lesamouraipourpre** [comment]
If I understand the following:
https://github.com/adafruit/circuitpython/blob/7df5d74d951f41347502087ef851c19e76c7a2f7/py/persistentcode.c#L291-L308
```c
STATIC qstr load_qstr(mp_reader_t *reader, qstr_window_t *qw) {
    size_t len = read_uint(reader, NULL);
    if (len == 0) {
        // static qstr
        return read_byte(reader);
    }
    if (len & 1) {
        // qstr in window
        return qstr_window_access(qw, len >> 1);
    }
    len >>= 1;
    char *str = m_new(char, len);
    read_bytes(reader, (byte *)str, len);
    qstr qst = qstr_from_strn(str, len);
    m_del(char, str, len);
    qstr_window_push(qw, qst);
    return qst;
}
```

If the len is 0, the next byte defines the string in a static lookup table in `makeqstrdata.py`:
https://github.com/adafruit/circuitpython/blob/7df5d74d951f41347502087ef851c19e76c7a2f7/py/makeqstrdata.py#L87-L253

If the len low bit is set, use `qstr_window_access(qw, len >> 1)`

If the len low bit is not set, eg. `\x16` read in the next `len >> 1` bytes and store it with `qstr_window_push`

I've not figured out the double `\x16` yet. It might be because of https://github.com/adafruit/circuitpython/blob/7df5d74d951f41347502087ef851c19e76c7a2f7/py/persistentcode.c#L419 but I can't be sure of it. I'm still trying to wrap my brain around it all.

**Comment by askpatrickw** [comment]
Since that PR for adafruit_seesaw was released, anything else left to do?

**Comment by askpatrickw** [comment]
I'm going to give this another test but it seemed like this was good to go ...

**Comment by askpatrickw** [comment]
@lesamouraipourpre  and @Neradoc  is there anything you feel still needs testing on this?

File: circup.py [hunk 1]
```diff
@@ -318,19 +359,29 @@ def extract_metadata(path):
...
-        # value of 11 to indicate length of "__version__").
-        loc = content.find(b"\x0b__version__")
+        # Track the MPY version number
+        mpy_version = content[0:2]
+        compatibility = None
+        # Find the start location of the __version__
+        if mpy_version == b"M\x03":
+            # One byte for the length of "__version__"
+            loc = content.find(b"__version__") - 1
+            compatibility = (None, "7.0.0-alpha.1")
```
Comment (askpatrickw) [inline circup.py:373]: Why is this specifically "7.0.0-alpha.1" ?

File: circup.py [hunk 2]
```diff
@@ -318,19 +359,29 @@ def extract_metadata(path):
...
+        elif mpy_version == b"C\x05":
+            # Two bytes in mpy version 5
+            loc = content.find(b"__version__") - 2
+            compatibility = ("7.0.0-alpha.1", None)
```
Comment (askpatrickw) [inline circup.py:377]: Why is this specifically "7.0.0-alpha.1" ?

**Review by askpatrickw — COMMENTED** [review]
Just one question the the specific reference to 7.0.0 alpha

File: circup.py (hunk 1 above)
Comment (lesamouraipourpre) [reply circup.py:373]: I think it's the first version that uses the new upstream mpy format.

**Comment by askpatrickw** [comment]
As you knew would happen... @Neradoc This now has merge conflicts, can you resolve them?

**Comment by lesamouraipourpre** [comment]
I believe it was good to go. It looks like a recent PR merge has introduced conflicts. Hopefully these are easy to resolve.

File: circup.py (hunk 1 above)
Comment (Neradoc) [reply circup.py:373]: Yep, it's the first tag that has the MPY change, not that previous versions of the alpha need support anyway, but it was a practical cut-off.

**Comment by Neradoc** [comment]
Rebased, quickly re-tested, seems good so far.

File: circup.py (hunk 1 above)
Comment (askpatrickw) [reply circup.py:465]: Got it... I didn't quite understand how it was being used. Re-read the code and understand now. Thanks for humoring me. :-)

**Review by askpatrickw — APPROVED** [review]
reviewed and lightly tested

**Pull request merged**